internal/services/cdn/cdn_frontdoor_helpers.go

```diff
@@ -339,76 +383,265 @@ func updateRouteAssociations(d *pluginsdk.ResourceData, meta interface{}, routeI
+		}
+
+		// Make sure all of the custom domains that are referenced by the resource are actually associated with the route...
+		notAssociated := make([]string, 0)
+
+		for _, v := range resourceCustomDomains {
+			customDomain, err := parse.FrontDoorCustomDomainIDInsensitively(v.(string))
+			if err != nil {
+				return fmt.Errorf("unable to parse %q: %+v", v.(string), err)
+			}
+
+			if !sliceContainsString(routeCustomDomains, customDomain.ID()) {
+				notAssociated = append(notAssociated, fmt.Sprintf("%q", customDomain.ID()))
+			}
+		}
+
+		if len(notAssociated) > 0 {
+			return fmt.Errorf("contains CDN FrontDoor Custom Domains that are not associated with the CDN FrontDoor Route(Name: %q). Please remove the following CDN FrontDoor Custom Domain(s) from your CDN Route Disable Link To Default Domain configuration: %s", routeName, strings.Join(notAssociated, ", "))
+		}
+	}
+
+	return nil
+}
+
+func validateRoutesCustomDomainProfile(customDomains []interface{}, routeName string, routeProfile string) error {
 	wrongProfile := make([]string, 0)
 
-	for _, v := range resourceCustomDomains {
-		customDomain, err := parse.FrontDoorCustomDomainIDInsensitively(v.(string))
-		if err != nil {
-			return err
+	if len(customDomains) != 0 {
+		// Verify all of the custom domains belong to the same profile as the route...
+		for _, v := range customDomains {
+			customDomain, err := parse.FrontDoorCustomDomainIDInsensitively(v.(string))
+			if err != nil {
+				return err
+			}
+
+			if customDomain.ProfileName != routeProfile {
+				wrongProfile = append(wrongProfile, fmt.Sprintf("%q", customDomain.ID()))
+			}
 		}
 
-		if customDomain.ProfileName != routeProfile {
-			wrongProfile = append(wrongProfile, fmt.Sprintf("%q", customDomain.ID()))
+		if len(wrongProfile) > 0 {
+			return fmt.Errorf("the following CDN FrontDoor Custom Domain(s) do not belong to the expected CDN FrontDoor Profile(Name: %q). Please remove the following CDN FrontDoor Custom Domain(s) from your CDN FrontDoor Route configuration block: %s", routeProfile, strings.Join(wrongProfile, ", "))
 		}
 	}
 
-	if len(wrongProfile) > 0 {
-		return fmt.Errorf("the following CDN Front Door Custom Domain(s) do not belong to the expected CDN Front Door Profile(Name: %q). Please remove the following CDN Front Door Custom Domain(s) from your CDN Route Disable Link To Default Domain configuration: %s", routeProfile, strings.Join(wrongProfile, ", "))
+	return nil
+}
+
+// Validates that the CDN FrontDoor Custom Domain can be associated with the CDN FrontDoor Route
+func validateCustomDomainRoutes(input *[]parse.FrontDoorRouteId, customDomainID *parse.FrontDoorCustomDomainId) error {
+	if len(*input) == 0 || input == nil {
```

Comment:
there's a crash here:

```suggestion
	if input == nil || len(*input) == 0 {
```